# Bài tập về nhà tuần 2




## Bài 1: Tự xây dựng hàm xử lý:

- a. Xây dựng 1 hàm trả về danh sách mảng gồm n phần tử. Trong đó, các phần tử nào bị duplicated với nhau được đánh dấu là True.

- b. Xây dựng 1 hàm trả về danh sách mảng chứa các vị trí dữ liệu bị missing value. 

- c. Xây dựng hàm kiểm tra về bất đồng bộ dữ liệu. 
    
    + **Mục tiêu chính:** Phát hiện được vấn đề chênh lệch kích thước, format 
    + **Nâng cao(cộng điểm):** Phát hiện kiểm tra xem có ký tự đặc biệt nào ngoài các ký tự có thể giá trị đó. Ví dụ phát hiện tồn tại A@1232, 113462A.

- d. Xây dựng hàm xử lý vấn đề bất đồng bộ

In [18]:
import pandas as pd
import numpy as np

In [129]:
data = pd.DataFrame(
    {
        'MSV':['A43471','40401','A39380','A38843','A41131','A30426','A30479','A33541','A30012','A31111','A42356','A43321','A35614','A43211','40401'],
         'DiemQT':[1,7,6,7,5,10,6,7,9,5,1,6,8,6,5],
        'GioiTinh':['Nam','Nam','Nu','Nam','Nu','Nam','Nam','Nam',0,1,1,0,1,0,5]
    }
)

# Bài 1

### a.Xây dựng 1 hàm trả về danh sách mảng gồm n phần tử. Trong đó, các phần tử nào bị duplicated với nhau được đánh dấu là True.

In [130]:
def duplicated(a):
    dup_a = np.zeros(len(a),dtype=bool)
    for i in range(len(a)-1):
        for j in range(i+1,len(a)):
            if i != j and a[i] == a[j]:
                dup_a[j] = True
                dup_a[i] = True
    return dup_a

In [131]:
data.duplicated()

0     False
1     False
2     False
3     False
4     False
5     False
6     False
7     False
8     False
9     False
10    False
11    False
12    False
13    False
14    False
dtype: bool

In [132]:
duplicated(data.MSV)

array([False,  True, False, False, False, False, False, False, False,
       False, False, False, False, False,  True])

In [133]:
duplicated(data.DiemQT)

array([ True,  True,  True,  True,  True, False,  True,  True, False,
        True,  True,  True, False,  True,  True])

In [134]:
duplicated(data.GioiTinh)

array([ True,  True,  True,  True,  True,  True,  True,  True,  True,
        True,  True,  True,  True,  True, False])

### b. Xây dựng 1 hàm trả về danh sách mảng chứa các vị trí dữ liệu bị missing value

In [135]:
def missing(n):
    a =[]
    for x in n:
        for i in range(len(n)):
            if n[x][i] == None or n[x][i] != n[x][i] :
                a = np.append(a,i)
    return a

In [136]:
missing(data)

[]

### c. Xây dựng hàm kiểm tra về bất đồng bộ dữ liệu.

Mục tiêu chính: Phát hiện được vấn đề chênh lệch kích thước, format

Nâng cao(cộng điểm): Phát hiện kiểm tra xem có ký tự đặc biệt nào ngoài các ký tự có thể giá trị đó. Ví dụ phát hiện tồn tại A@1232, 113462A.

In [149]:
def batdong(a):
    
    bd = np.ones(len(a), dtype= bool)
    for i in range(len(a)-1):
        khac = 0
        cung = 0
        for j in range(i+1,len(a)):
            if type(a[i]) != type(a[j]) or len(a[i]) != len(a[j]):
                   khac += 1
            else: 
                   cung += 1
        if khac > cung:
            bd[i] = False
    return bd
                   

In [150]:
batdong(data.MSV)

array([ True, False,  True,  True,  True,  True,  True,  True,  True,
        True,  True,  True,  True, False,  True])

### d. Xây dựng hàm xử lý vấn đề bất đồng bộ

In [ ]:
def xuli(n)
    a=[]
    l=0
    for x in n:
        for i in range(len(n)):
            if (len(n[x][i] < l):
                n[x][i] 

In [153]:
for i in range (len(data)-1):
    if (len(data['MSV'][i]) < 6): 
        if "A" not in data['MSV'][i]:
            data['MSV'][i] = 'A' + data['MSV'][i]

C:\Users\Admin\AppData\Local\Temp\ipykernel_6024\2951563593.py:4: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  data['MSV'][i] = 'A' + data['MSV'][i]




## Bài 2

Dữ liệu mô tả về số lượng dân số của các nước trên toàn thế giới từ năm 1960-2021. Thực hiện một số yêu cầu sau:

+ Phát hiện các vấn đề của tập dữ liệu. Lưu ý, dùng Markdown để diễn giải các lỗi gặp phải.
+ Đề xuất phương pháp xử lý với các vấn đề đó. Dùng markdown giải thích phương pháp xử lý, nguyên do lựa chọn phương pháp đó.
+ Sử dụng hàm ở Bài 1 để thử nghiệm với dữ liệu bài 2

In [123]:
%cd "C:\Users\Admin\Documents\Nhập môn khoa học dữ liệu\W2"

C:\Users\Admin\Documents\Nhập môn khoa học dữ liệu\W2


In [124]:
df = pd.read_csv('data_btvn.csv')

In [146]:
df

Country Name Country Code         1960         1961  \
0                          Aruba          ABW      54208.0      55434.0   
1    Africa Eastern and Southern          AFE  130836765.0  134159786.0   
2                    Afghanistan          AFG    8996967.0    9169406.0   
3     Africa Western and Central          AFW   96396419.0   98407221.0   
4                         Angola          AGO    5454938.0    5531451.0   
..                           ...          ...          ...          ...   
261                       Kosovo          XKX     947000.0     966000.0   
262                  Yemen, Rep.          YEM    5315351.0    5393034.0   
263                 South Africa          ZAF   17099836.0   17524533.0   
264                       Zambia          ZMB    3070780.0    3164330.0   
265                     Zimbabwe          ZWE    3776679.0    3905038.0   

            1962         1963         1964         1965         1966  \
0        56234.0      56699.0      57029.0      57357.0      57702.0   
1    137614644.0  141202036.0  144920186.0  148769974.0  152752671.0   
2      9351442.0    9543200.0    9744772.0    9956318.0   10174840.0   
3    100506960.0  102691339.0  104953470.0  107289875.0  109701811.0   
4      5608499.0    5679409.0    5734995.0    5770573.0    5781305.0   
..           ...          ...          ...          ...          ...   
261     994000.0    1022000.0    1050000.0    1078000.0    1106000.0   
262    5473671.0    5556767.0    5641598.0    5727745.0    5816241.0   
263   17965733.0   18423157.0   18896303.0   19384838.0   19888259.0   
264    3260645.0    3360099.0    3463211.0    3570466.0    3681953.0   
265    4039209.0    4178726.0    4322854.0    4471178.0    4623340.0   

            1967  ...         2012         2013         2014         2015  \
0        58044.0  ...     102565.0     103165.0     103776.0     104339.0   
1    156876454.0  ...  547482863.0  562601578.0  578075373.0  593871847.0   
2     10399936.0  ...   31161378.0   32269592.0   33370804.0   34413603.0   
3    112195950.0  ...  370243017.0  380437896.0  390882979.0  401586651.0   
4      5774440.0  ...   25107925.0   26015786.0   26941773.0   27884380.0   
..           ...  ...          ...          ...          ...          ...   
261    1135000.0  ...    1807106.0    1818117.0    1812771.0    1788196.0   
262    5907873.0  ...   24473176.0   25147112.0   25823488.0   26497881.0   
263   20406863.0  ...   52832659.0   53687125.0   54544184.0   55386369.0   
264    3797877.0  ...   14465148.0   14926551.0   15399793.0   15879370.0   
265    4779825.0  ...   13115149.0   13350378.0   13586710.0   13814642.0   

            2016         2017         2018         2019         2020  \
0       104865.0     105361.0     105846.0     106310.0     106766.0   
1    609978946.0  626392880.0  643090131.0  660046272.0  677243299.0   
2     35383028.0   36296111.0   37171922.0   38041757.0   38928341.0   
3    412551299.0  423769930.0  435229381.0  446911598.0  458803476.0   
4     28842482.0   29816769.0   30809787.0   31825299.0   32866268.0   
..           ...          ...          ...          ...          ...   
261    1777557.0    1791003.0    1797085.0    1788878.0    1790133.0   
262   27168210.0   27834811.0   28498683.0   29161922.0   29825968.0   
263   56207649.0   57009751.0   57792520.0   58558267.0   59308690.0   
264   16363449.0   16853608.0   17351714.0   17861034.0   18383956.0   
265   14030338.0   14236599.0   14438812.0   14645473.0   14862927.0   

            2021  
0       107195.0  
1    694665117.0  
2     39835428.0  
3    470898870.0  
4     33933611.0  
..           ...  
261    1806279.0  
262   30490639.0  
263   60041996.0  
264   18920657.0  
265   15092171.0  

[266 rows x 64 columns]

In [126]:
duplicated(df['Country Name']).sum()

0

In [127]:
duplicated(df['Country Code']).sum()

0

- Không bị missing value

In [128]:
pd.DataFrame(missing(df)).value_counts()

110.0    62
196.0    30
69.0     10
127.0     3
dtype: int64

- Tại vị trí 110 bị trống 62/64 cột 

    Giải pháp: Xoá cột ấy đi> 

    Nguyên do: Vì bị trống quá 50%
    

- Tại vị trí 196 bị trống 30/64 cột 

    Giải pháp: Xoá cột đi

    Nguyên do: Vì bị trống quá 50%
    

- Tại vị trí 69 bị trống 10/64 cột

    Giải pháp: Thêm mean 

    Nguyên do: Vì bị trống ít
    

- Tại vị trí 127 bị trống 3/64 cột

    Giải pháp: Thêm mean

    Nguyên do: Vì bị trống ít

In [151]:
batdong(df['Country Name']).sum()

1

Không đồng bộ dữ liệu Country Name